X-Ray -> OTel traces conversion with improved observability

hencrice opened the pull request:
**Description:**
Implemented the conversion from X-Ray segment to Otel trace format. The observability of the internals of this receiver was enhanced with `obsreport`.

**Link to tracking Issue:** <Issue number if applicable>
#585
#478

**Testing:**
Existing unit tests updated. Many new ones added

**Documentation:** < Describe the documentation added.>
N/A

hencrice committed:
X-Ray -> OT traces conversion with improved observability

The observability of the internals of this receiver was enhanced with `obsreport`.
Closes #585. Closes #478.

hencrice committed:
Made the sample programs compile with makefile target

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion

hencrice committed:
Added copyright headers

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion

hencrice committed:
Changed import order

hencrice committed:
Minor changes to re-trigger load test, which failed on the previous revision. Ironically, my changes are not even part of the binary that was being load tested.

anuraaga commented on receiver/awsxrayreceiver/internal/translator/addtoattrs.go:
```diff
@@ -0,0 +1,51 @@
+
+const (
+	// AWSXRayInProgressAttribute is the `in_progress` flag in an X-Ray segment
+	AWSXRayInProgressAttribute = "aws.xray.inprogress"
```
Similary, I think we'll need custom x-ray attributes to be shared between receiver and exporter and should go in the shared package

anuraaga commented on receiver/awsxrayreceiver/internal/translator/aws.go:
```diff
@@ -0,0 +1,71 @@
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+
+	expTrans "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsxrayexporter/translator"
```
Same theme, let's start with the extraction

anuraaga commented on receiver/awsxrayreceiver/internal/translator/cause.go:
```diff
@@ -0,0 +1,102 @@
+	}
+}
+
+func convertStackFramesToStackTraceStr(stack []tracesegment.StackFrame) string {
```
Let's go ahead and construct a Java stack trace (added the parsing code yesterday) instead of a custom stacktrace format. We can use that for the causes too I believe to avoid the ExceptionID issue

anuraaga commented on receiver/awsxrayreceiver/internal/translator/http.go:
```diff
@@ -0,0 +1,64 @@
+			// since the ClientIP is not nil, this means that this segment is generated
+			// by a server serving an incoming request
+			attrs.UpsertString(conventions.AttributeHTTPClientIP, *req.ClientIP)
+			span.SetKind(pdata.SpanKindSERVER)
```
HTTP doesn't seem like the right file to be setting the kind. Anyways, isn't it just whether it's a segment or a subsegment and whether it's remote or not?

hencrice replied:
Not sure I understand what you were getting at here.

The main reason I put it here is because the `ClientIP` is the sure way to determine whether this is a server span or client span:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#spankind

anuraaga replied:
Well we're setting kind in `name.go`. It's too mysterious to also have kind determined in `http.go`, especially since it's not really a feature of HTTP. I think it's ok to peek into the HTTP struct again in a different package and is a win to keep the logic for determining kind together.

For client_ip, I was worried because the data model docs say all fields of HTTP are optional - determining one of the most important fields based on an optional is risky. In practice, I think all our SDKs do set it though so it may be ok. But if we just set kind server for all segments I think it's ok too - while it's possible to have `INTERNAL` segments, none of the other data models support this. So in the xray exporter we just map all server spans to segments

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/segment.go#L201

hencrice replied:
I moved 1 line to `name.go`. For your suggestion on setting default kind to `SpanKindSERVER` for all segments, the main problem is I can not tell when I should set kind to `pdata.SpanKindCLIENT`. Essentially, I don't know what to write for XXX:

```
if XXX is true:
    span.SetKind(pdata.SpanKindCLIENT)
else:
    span.SetKind(pdata.SpanKindSERVER)
```

which is why I wrote it using the optional `ClientIP` field:
```
if seg.HTTP != nil && seg.HTTP.Request != nil && seg.HTTP.Request.ClientIP != nil {
	// since the ClientIP is not nil, this means that this segment is generated
	// by a server serving an incoming request
	span.SetKind(pdata.SpanKindSERVER)
} else {
	span.SetKind(pdata.SpanKindCLIENT)
}
```

hencrice replied:
I have updated the code but can't publish another revision yet since it also contains changes in #702. Still waiting on that PR.

anuraaga commented on receiver/awsxrayreceiver/internal/translator/http.go:
```diff
@@ -0,0 +1,64 @@
+		if resp.Status != nil {
+			otStatus := tracetranslator.OCStatusCodeFromHTTP(int32(*resp.Status))
+			// in X-Ray exporter, the segment status is set:
+			// first via the span attribute, conventions.AttributeHTTPStatusCode
```
Don't know if this is right, or maybe I'm misunderstanding. The segment status is based on the otel status (which is translated to http status)

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/cause.go#L147

But otel status is going away anyways so not a big deal

anuraaga commented on receiver/awsxrayreceiver/internal/translator/parentspanid.go:
```diff
@@ -0,0 +1,34 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver/internal/tracesegment"
+)
+
+func addParentSpanID(seg *tracesegment.Segment, parentID *string, span *pdata.Span) {
```
This helper doesn't seem very helpful, the function signature doesn't make clear what it does. If you keep it this needs some detailed godoc

anuraaga commented on receiver/awsxrayreceiver/internal/translator/sdk.go:
```diff
@@ -0,0 +1,31 @@
+)
+
+func populateInstrumentationLibrary(seg *tracesegment.Segment, ils *pdata.InstrumentationLibrarySpans) {
+	if seg.AWS != nil && seg.AWS.XRay != nil {
```
This should be `Resource.sdk.name` / language / version I think. Unfortunately we'll need to parse into a name and language. The exporter ignores instrumentationlibrary so you can fill it with any dummy value

hencrice replied:
Gonna add these information to both the `instrumentationlibrary` and as resource attributes

anuraaga commented on receiver/awsxrayreceiver/internal/translator/sql.go:
```diff
@@ -0,0 +1,65 @@
+		return nil
+	}
+
+	// https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/sql.go#L33
```
Less important than having code and attributes in a shared location, so this can come later, but the translation in the common location might be nice too eventually, a class with `fromPdata` and `toPdata` in the same location is very easy to reason about instead of jumping around.

hencrice replied:
Added a TODO

anuraaga commented on receiver/awsxrayreceiver/internal/translator/sql.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+func splitSQLURL(rawURL string) (string, string, error) {
+	m := re.FindStringSubmatch(rawURL)
```
I think you can just use `strings.LastIndexByte(rawURL, '/')` instead of the regex

hencrice replied:
I did that but realized that for an URL looks like this: `"jdbc:postgresql://aawijb5u25wdoy.cpamxznpdoq8.us-west-2.rds.amazonaws.com:5432/ebdb"`, if customers forget to provide `/ebdb` part, then I will treat `aawijb5u25wdoy.cpamxznpdoq8.us-west-2.rds.amazonaws.com:5432` as the db name and `jdbc:postgresql:/` as URL, which is incorrect.

anuraaga commented on receiver/awsxrayreceiver/internal/translator/translator.go:
```diff
@@ -0,0 +1,190 @@
+
+	addBool(seg.Traced, AWSXRayTracedAttribute, &attrs)
+
+	// TODO: transform the `metadata` field once X-Ray exporter
```
x-ray exporter will deal with metadata differently for OTel instrumented apps than this receiver I think. There is no concept of metadata / annotation in OTel, so we'll use key lists to control which it becomes (/cc @shengxil) since prefixing is complicated and breaks compliance with other backends.

For this receiver, we should instead define a convention prefix, maybe `opentelemetry.internal.collector.<metadata key> = JSON.serialize(<metadata value>)` and process that specially in the exporter.

hencrice replied:
As you mentioned, there's no concept in OT between attributes that will be indexed (i.e. X-Ray annotations) and those that will not be (X-Ray metadata). So for now I will go ahead and define a prefix that will only be used by the X-Ray OT receiver and exporter

anuraaga commented on receiver/awsxrayreceiver/internal/translator/translator.go:
```diff
@@ -0,0 +1,190 @@
+)
+
+const (
+	// just a guess to avoid too many memory re-allocation
```
Unless this guess is x-ray specific I'd leave it out and use default, otherwise document what the guess is based on

hencrice replied:
It's not x-ray specific. However, there's no default for https://pkg.go.dev/go.opentelemetry.io/collector@v0.7.0/consumer/pdata?tab=doc#AttributeMap. Either we just initialize with size 0 (and potentially experience reallocation a few times when we insert number of attributes) or a non-zero guess to avoid re-allocation before we have 15 attributes.

anuraaga reviewed (commented):
Thanks, mostly looks good but recommend waiting on this PR and doing one preparation one first by extracting common bits from the exporter

hencrice committed:
WIP. Addressed some feedbacks. Still need to do metadata and construct java stacktrace

hencrice committed:
Created shared x-ray segment model

Also migrated the X-Ray exporter to use this new data model. This new model will also be used in #691.

This PR closes #699.

hencrice committed:
Need to move util over here

hencrice committed:
Updated receiver and all tests

hencrice committed:
Addressed feedbacks

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray/datamodel

hencrice committed:
Fixed broken tests

hencrice committed:
Merge branch 'x-ray/datamodel' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion_contDev

Need to update existing code in the receiver to use the new shared data
model

hencrice committed:
Updated the receiver to use the shared X-Ray segment data model

hencrice committed:
Need to remove the awsxray/util package and revert back to aws

hencrice committed:
Removed the util package

hencrice committed:
Merge branch 'x-ray/datamodel' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion_contDev

hencrice committed:
Removed the use of the awsxray/util package

hencrice committed:
Handle the metadata fields

hencrice committed:
Re-trigger the load test

The test does not seem to be stable enough.

hencrice commented:
I have all the changes required to address the feedbacks ready to be reviewed pushed to a separate branch. Waiting on #702 to be merged first before updating this PR.

hencrice committed:
Addressed feedbacks on setting spanKind in http.go

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray/datamodel

hencrice committed:
Merge branch 'x-ray/datamodel' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion_contDev

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray/datamodel

hencrice committed:
Merge branch 'x-ray/datamodel' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion_contDev

hencrice committed:
Fixed typo

bogdandrutu commented:
Does OT means Otel? If so call it that way

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion_contDev

hencrice commented:
> Does OT means Otel? If so call it that way

Renamed the PR and description

hencrice committed:
Missed one merge conflict in x-ray exporter

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion

anuraaga commented on receiver/awsxrayreceiver/internal/translator/aws.go:
```diff
@@ -0,0 +1,70 @@
+
+func addAWSToResource(aws *awsxray.AWSData, attrs *pdata.AttributeMap) {
+	if aws == nil {
+		// https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/aws.go#L153
```
Either pemalink or a description of the problem without the link, line link easy to drift (it already did)

anuraaga commented on receiver/awsxrayreceiver/internal/translator/aws.go:
```diff
@@ -0,0 +1,70 @@
+		// https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/aws.go#L153
+		// this implies that the current segment being processed is not generated
+		// by an AWS entity.
+		attrs.UpsertString(conventions.AttributeCloudProvider, "nonAWS")
```
```suggestion
		attrs.UpsertString(conventions.AttributeCloudProvider, "unknown")
```

anuraaga commented on receiver/awsxrayreceiver/internal/translator/aws.go:
```diff
@@ -0,0 +1,70 @@
+		addString(aws.RequestID, awsxray.AWSRequestIDAttribute, attrs)
+		addString(aws.QueueURL, awsxray.AWSQueueURLAttribute, attrs)
+		addString(aws.TableName, awsxray.AWSTableNameAttribute, attrs)
+		// the "retries" field is dropped
```
We can define more AWS-specific attributes if needed let's avoid dropping data by following up in another PR

anuraaga commented on receiver/awsxrayreceiver/internal/translator/cause.go:
```diff
@@ -0,0 +1,114 @@
+				attrs.UpsertString(conventions.AttributeExceptionStacktrace, stackTrace)
+				// For now X-Ray's exception data model is not fully supported in the OpenTelemetry
+				// spec, so some information is lost here.
+				// For example, the "cause" ,"remote", ... and some fields within each exception
```
I think cause is very important so if not in this PR we need to follow up very quickly.

It's grungy, but can you merge the causes into a single exception event with a full stack trace?

hencrice replied:
1. Will add code to send all those as attribute as custom AWS attributes. However, the X-Ray exporter currently does not handle those, so it needs to be updated as well

2. Just want to make sure I understand you, you want to merge multiple exceptions into 1 with a bigger stacktrace. X-Ray exporter can handle multiple exception events: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c615d2db351929b99e46f7b427f39c12afe15b54/exporter/awsxrayexporter/translator/cause.go#L61, so I don't think we should merge it.

anuraaga commented on receiver/awsxrayreceiver/internal/translator/cause.go:
```diff
@@ -0,0 +1,114 @@
+)
+
+const (
+	newLine   = "\n"
```
While there's some maintenance advantage to having constants for these, I think we'll get much more readability by inlining them below

anuraaga commented on receiver/awsxrayreceiver/internal/translator/cause.go:
```diff
@@ -0,0 +1,114 @@
+	for _, frame := range excp.Stack {
+		line := strconv.Itoa(*frame.Line)
+		// the string representation of a frame looks like:
+		// <*frame.Label>\n<*frame.Path>:line\n
```
```suggestion
		// <*frame.Label>(<*frame.Path>):line\n
```

anuraaga commented on receiver/awsxrayreceiver/internal/translator/http.go:
```diff
@@ -0,0 +1,63 @@
+)
+
+func addHTTP(seg *awsxray.Segment, span *pdata.Span) {
+	span.Status().InitEmpty() // by default this sets the code to `Status_Ok`
```
Can you initialize this at a higher level? `addHTTP` seems too low to initialize the status

anuraaga commented on receiver/awsxrayreceiver/internal/translator/annotations.go:
```diff
@@ -0,0 +1,44 @@
+		case float64:
+			attrs.UpsertDouble(k, t)
+		default:
+			attrs.UpsertString(k, fmt.Sprintf("%+v", t))
```
Is this possible for annotations? If we don't expect it let's just ignore instead of hiding it with this

hencrice replied:
I believe so. `annos` is of type `map[string]interface{}` and I can't tell whether interface can be an object.

anuraaga commented on receiver/awsxrayreceiver/internal/translator/sdk.go:
```diff
@@ -0,0 +1,55 @@
+func populateInstrumentationLibrary(seg *awsxray.Segment, ils *pdata.InstrumentationLibrarySpans) {
+	if seg.AWS != nil && seg.AWS.XRay != nil {
+		xr := seg.AWS.XRay
+		il := ils.InstrumentationLibrary()
```
We don't use instrumentation library in exporter, and even if we did it would be for something different than the SDK version - x-ray doesn't propagate something semantically similar to instrumentation library. So let's leave this out.

anuraaga commented on receiver/awsxrayreceiver/internal/translator/sdk.go:
```diff
@@ -0,0 +1,55 @@
+				// TODO: Update this once the exporter is more flexible
+				attrs.UpsertString(conventions.AttributeTelemetrySDKLanguage, "java")
+			} else {
+				res := strings.Split(*xr.SDK, "for ")
```
Use index and slice instead of split

hencrice replied:
Not exactly sure what you meant here, so I switched to use `strings.Index`

anuraaga commented on receiver/awsxrayreceiver/internal/translator/sql.go:
```diff
@@ -0,0 +1,65 @@
+	return nil
+}
+
+var re = regexp.MustCompile(`^(.+\/\/.+)\/(.+)$`)
```
Can you add a comment for this regex?

anuraaga commented on receiver/awsxrayreceiver/internal/translator/translator.go:
```diff
@@ -0,0 +1,192 @@
+		if seg.Cause != nil &&
+			populatedChildSpan.Status().Code() != pdata.StatusCode(otlptrace.Status_Ok) {
+			// if seg.Cause is not nil, then one of the subsegments must contain a
+			// HTTP error code. ALso, span.Status().Code() is already
```
```suggestion
			// HTTP error code. Also, span.Status().Code() is already
```

anuraaga commented on receiver/awsxrayreceiver/internal/translator/translator.go:
```diff
@@ -0,0 +1,192 @@
+			// HTTP error code. ALso, span.Status().Code() is already
+			// set to `otlptrace.Status_UnknownError` by `addCause()` in
+			// `populateSpan()` above, so here we are just trying to figure out
+			// whether we can get a event more specific error code.
```
```suggestion
			// whether we can get an even more specific error code.
```

anuraaga reviewed (commented):
Just some small comments

hencrice committed:
Apply suggestions from code review

More changes to follow up

Co-authored-by: Anuraag Agrawal <anuraaga@gmail.com>

hencrice commented on receiver/awsxrayreceiver/go.mod:
```diff
@@ -5,13 +5,11 @@ go 1.14
+	github.com/open-telemetry/opentelemetry-proto v0.4.0
 	github.com/stretchr/testify v1.6.1
-	go.opentelemetry.io/collector v0.8.1-0.20200819173546-64befbcc0564
+	go.opentelemetry.io/collector v0.8.0
```
I think it's because I tend to delete packages listed in go.mod and re-run `go test` whenever there's a merge conflict. And on go package registry, the latest collector version is 0.8.0

hencrice committed:
Addressed the remaining comments

anuraaga commented on receiver/awsxrayreceiver/internal/translator/cause.go:
```diff
@@ -0,0 +1,112 @@
+		exceptionEventStartIndex := evts.Len()
+		evts.Resize(exceptionEventStartIndex + len(seg.Cause.Exceptions))
+
+		for i, excp := range seg.Cause.Exceptions {
```
Again fine as a followup but we should merge exceptions in the cause chain

- It's how OTel reports exceptions, we want to report the same. AWS Exporter currently expects all exceptions in the cause chain to be in the same stack trace and I'm not sure how it will deal with having them as multiple events.

- If users use a different exporter it definitely won't work properly with the cause chain broken up. Our custom attributes are a fallback for fidelity with xray, but we still want to follow the OTel format as much as possible to interop well with the other exporters.

anuraaga commented on receiver/awsxrayreceiver/internal/translator/sql.go:
```diff
@@ -0,0 +1,66 @@
+	return nil
+}
+
+// SQL URL is of the format: protocol+transport://host:port/dbName
```
Thanks for the comment - it seems like the regex won't handle query parameters in th eURL like many JDBC urls? `mysql://localhost:1234/mydb?myInterceptor=foo`

anuraaga reviewed (approved):
Thanks - we can follow up on the point about causes in another PR but do look into the regex to make sure it's ok. Approving since other than that LGTM

hencrice committed:
Addressed feedback on sql.go

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion

hencrice committed:
Set up environment variables

hencrice committed:
Fixed regex and added tests

anuraaga reviewed (approved):
Thanks!

hencrice commented:
@bogdandrutu Could you take a look and merge it? Thanks!

bogdandrutu commented on receiver/awsxrayreceiver/go.mod:
```diff
@@ -7,6 +7,8 @@ replace github.com/open-telemetry/opentelemetry-collector-contrib/internal/commo
 require (
 	github.com/aws/aws-sdk-go v1.34.8
 	github.com/google/uuid v1.1.1
+	github.com/open-telemetry/opentelemetry-collector-contrib/internal/common v0.0.0-00010101000000-000000000000
```
Need to have a replace with local here

hencrice replied:
oh that was done. It's on ln5. Will try to split common/aws in its own package

bogdandrutu reviewed (commented):
Consider to split common/AWS in its own package

hencrice committed:
Split internal/common/awsxray into its own package

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/otlpConversion

hencrice committed:
Add Makefile
